### PR TITLE
Request end callback

### PR DESCRIPTION
**Pull request by bryanmau1**
R: @andrewsg @jcollins-g

### CODE REVIEW

**Comment by bryanmau1** [comment]
I thought I registered my GitHub account hmmm

**Comment by bryanmau1** [comment]
Just FYI, this PR does not fix the NDB bug.  It just implements the callback feature for this runtime, a necessary prerequisite for the NDB fix.

File: appengine-vmruntime/vmruntime/middleware.py [hunk 1]
```diff
@@ -235,3 +236,22 @@ def health_check_intercept_wrapper(request):
...
+
+    Returns:
+        The wrapped app, also a WSGI app.
+    """
+    def callback_wrapper(environ, start_response):
+        """Call the WSGI app and then invoke the request-end callbacks."""
+        try:
+            return app(environ, start_response)
+        finally:
+            callback.InvokeCallbacks()
```
Comment (andrewsg) [inline appengine-vmruntime/vmruntime/middleware.py]: Hmm, I'm not sure if there might be any strange consequences to doing it with a finally:. I was imagining you would simply execute the callback as the final step before returning the response -- for instance,

    response = app(environ, start_response)
    callback.InvokeCallbacks()
    return response

I'm not sure about the pros and cons yet, so I'm okay with whichever method you choose. LGTM.

**Comment by andrewsg** [comment]
LGTM, after you've decided either way regarding the above note.

**Comment by bryanmau1** [comment]
Though WSGI applications are not supposed to throw errors and instead simply return error responses, it is possible and in some cases even intended.  So, try - finally will execute callbacks even in such situations.

**Comment by andrewsg** [comment]
Good thinking!

**Pull request merged**